resolver/resolver-https.go

```diff
@@ -0,0 +1,140 @@
+package resolver
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"github.com/miekg/dns"
+)
+
+// HttpsResolver is a resolver using just a single tcp connection with pipelining.
+type HttpsResolver struct {
+	BasicResolverConn
+	Client *http.Client
+}
+
+// HttpsQuery holds the query information for a httpsResolverConn.
+type HttpsQuery struct {
+	Query    *Query
+	Response chan *dns.Msg
+}
+
+// MakeCacheRecord creates an RRCache record from a reply.
+func (tq *HttpsQuery) MakeCacheRecord(reply *dns.Msg, resolverInfo *ResolverInfo) *RRCache {
+	return &RRCache{
+		Domain:   tq.Query.FQDN,
+		Question: tq.Query.QType,
+		RCode:    reply.Rcode,
+		Answer:   reply.Answer,
+		Ns:       reply.Ns,
+		Extra:    reply.Extra,
+		Resolver: resolverInfo.Copy(),
+	}
+}
+
+// NewHTTPSResolver returns a new HttpsResolver.
+func NewHTTPSResolver(resolver *Resolver) *HttpsResolver {
+	tr := &http.Transport{}
+
+	if resolver.ServerAddress != "" {
+		tr = &http.Transport{
+			TLSClientConfig: &tls.Config{
+				MinVersion: tls.VersionTLS12,
+				ServerName: resolver.VerifyDomain,
+				// TODO: use portbase rng
+			},
+		}
+	}
+
+	client := &http.Client{Transport: tr}
+
+	newResolver := &HttpsResolver{
+		BasicResolverConn: BasicResolverConn{
+			resolver: resolver,
+		},
+		Client: client,
+	}
+	newResolver.BasicResolverConn.init()
+	return newResolver
+}
+
+// Query executes the given query against the resolver.
+func (hr *HttpsResolver) Query(ctx context.Context, q *Query) (*RRCache, error) {
+
+	// Do not resolve domain names that are needed to initialize a resolver
+	if hr.resolver.Info.IP == nil {
+		if _, ok := resolverInitDomains[q.FQDN[:len(q.FQDN)-1]]; ok {
```

Comment:
Instead of tricking around here, please save the fully qualified domain name (ie. including the trailing `.`) in `resolverInitDomains`.
You can use `dns.Fqdn` for that.